### ISSUE TITLE
SNOW-636903: insert many rows into snowflake using subquery, it will throw error: "Bind variable ? not set"

### ISSUE THREAD

**Issue by xzxqq1** [snowflakedb/snowflake-connector-python#1211]
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

   Python 3.7.13 (default, Mar 28 2022, 07:24:34) 
   [Clang 12.0.0 ]

2. What operating system and processor architecture are you using?

   Darwin-21.1.0-x86_64-i386-64bit

3. What are the component versions in the environment (`pip freeze`)?

   asn1crypto==1.5.1
   certifi @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_83242e7e-f82d-4a71-8ef2- 
   9d71d212d249gu_wxmeq/croots/recipe/certifi_1655968827803/work/certifi
   cffi @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/croot-2p6q7e2r/cffi_1642701115063/work
   charset-normalizer==2.0.12
   cryptography==36.0.2
   idna==3.3
   oscrypto==1.3.0
   pycparser @ file:///tmp/build/80754af9/pycparser_1636541352034/work
   pycryptodomex==3.15.0
   PyJWT==2.4.0
   PyMySQL @ file:///opt/concourse/worker/volumes/live/bb2ed9f0-4942-4298-76ba- 
   b02a97cb1970/volume/pymysql_1610482900108/work
   pyOpenSSL==22.0.0
   pytz==2022.1
   requests==2.28.1
   snowflake-connector-python==2.7.9
   urllib3==1.26.11

4. What did you do?

   use snowflake python connector insert many row at once

snowflake_conn = snowflake.connector.connect(
    ...
    paramstyle='qmark'
)

snowflake_cursor = snowflake_conn.cursor()
try:
    # 'CLIENT_STAGE_ARRAY_BINDING_THRESHOLD': 65280
    # 6 flieds and 11000 rows at once
    snowflake_cursor.executemany('insert into xxx select column1, column2, column3, column4, column5, column6 from values (?, ?, ?, ?, ?, ?)', params)
finally:
    snowflake_cursor.close() 

5. What did you expect to see?

   It should inserted, but throw error

6. Can you set logging to DEBUG and collect the logs?

2022-07-29 00:23:41,767 - ThreadPoolExecutor-4_0 storage_client.py:189 - preprocess() - DEBUG - Preprocessing 11000.csv
2022-07-29 00:23:41,769 - ThreadPoolExecutor-4_0 network.py:1138 - _use_requests_session() - DEBUG - Session status for SessionPool 'b'sfc-va3-ds1-2-customer-stage.s3.amazonaws.com'', SessionPool 1/1 active sessions
2022-07-29 00:23:41,769 - ThreadPoolExecutor-4_0 storage_client.py:277 - _send_request_with_retry() - DEBUG - storage client request with session <snowflake.connector.vendored.requests.sessions.Session object at 0x7fc0fa763990>
2022-07-29 00:23:42,076 - ThreadPoolExecutor-4_0 connectionpool.py:461 - _make_request() - DEBUG - https://sfc-va3-ds1-2-customer-stage.s3.amazonaws.com:443 "HEAD /43j00000-s/stages/be6067a2-eef0-4f42-bce2-45b3ddbc7cbd/3f336c0a56374b6cb169ea40208d739c/11000.csv.gz HTTP/1.1" 404 0
2022-07-29 00:23:42,077 - ThreadPoolExecutor-4_0 network.py:1144 - _use_requests_session() - DEBUG - Session status for SessionPool 'b'sfc-va3-ds1-2-customer-stage.s3.amazonaws.com'', SessionPool 0/1 active sessions
2022-07-29 00:23:42,077 - ThreadPoolExecutor-4_0 s3_storage_client.py:403 - get_file_header() - DEBUG - not found. bucket: sfc-va3-ds1-2-customer-stage, path: 43j00000-s/stages/be6067a2-eef0-4f42-bce2-45b3ddbc7cbd/3f336c0a56374b6cb169ea40208d739c/11000.csv.gz
2022-07-29 00:23:42,077 - ThreadPoolExecutor-4_0 storage_client.py:118 - compress() - DEBUG - compressing file=11000.csv
2022-07-29 00:23:42,162 - ThreadPoolExecutor-4_0 storage_client.py:136 - get_digest() - DEBUG - getting digest file=11000.csv
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 file_util.py:145 - get_digest_and_size_for_stream() - DEBUG - getting digest and size for stream: cGi735b7lM3CYmBWdfGp6qWaiHlLwt7LeeERcypr67U=, 89280
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 storage_client.py:219 - prepare_upload() - DEBUG - Preparing to upload 11000.csv
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 storage_client.py:152 - encrypt() - DEBUG - encrypting file=11000.csv
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 encryption_util.py:76 - encrypt_stream() - DEBUG - key_size = 16
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 storage_client.py:225 - prepare_upload() - DEBUG - finished preprocessing
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 storage_client.py:230 - prepare_upload() - DEBUG - number of chunks 1
2022-07-29 00:23:42,163 - ThreadPoolExecutor-4_0 file_transfer_agent.py:450 - preprocess_done_cb() - DEBUG - Finished preparing file 11000.csv
2022-07-29 00:23:42,163 - ThreadPoolExecutor-3_0 storage_client.py:404 - upload_chunk() - DEBUG - Uploading chunk 0 of file 11000.csv
2022-07-29 00:23:42,164 - ThreadPoolExecutor-3_0 network.py:1138 - _use_requests_session() - DEBUG - Session status for SessionPool 'b'sfc-va3-ds1-2-customer-stage.s3.amazonaws.com'', SessionPool 1/1 active sessions
2022-07-29 00:23:42,164 - ThreadPoolExecutor-3_0 storage_client.py:277 - _send_request_with_retry() - DEBUG - storage client request with session <snowflake.connector.vendored.requests.sessions.Session object at 0x7fc0fa763990>
2022-07-29 00:23:42,856 - ThreadPoolExecutor-3_0 connectionpool.py:461 - _make_request() - DEBUG - https://sfc-va3-ds1-2-customer-stage.s3.amazonaws.com:443 "PUT /43j00000-s/stages/be6067a2-eef0-4f42-bce2-45b3ddbc7cbd/3f336c0a56374b6cb169ea40208d739c/11000.csv.gz HTTP/1.1" 200 0
2022-07-29 00:23:42,857 - ThreadPoolExecutor-3_0 network.py:1144 - _use_requests_session() - DEBUG - Session status for SessionPool 'b'sfc-va3-ds1-2-customer-stage.s3.amazonaws.com'', SessionPool 0/1 active sessions
2022-07-29 00:23:42,857 - ThreadPoolExecutor-3_0 storage_client.py:406 - upload_chunk() - DEBUG - Successfully uploaded chunk 0 of file 11000.csv
2022-07-29 00:23:42,857 - ThreadPoolExecutor-3_0 file_transfer_agent.py:497 - transfer_done_cb() - DEBUG - Chunk 0/1 of file 11000.csv reached callback
2022-07-29 00:23:42,857 - ThreadPoolExecutor-3_0 file_transfer_agent.py:513 - transfer_done_cb() - DEBUG - Chunk progress: 11000.csv: completed: 1 failed: 0 total: 1
2022-07-29 00:23:42,858 - ThreadPoolExecutor-3_0 storage_client.py:246 - finish_upload() - DEBUG - 11000.csv upload is completed.
2022-07-29 00:23:42,858 - ThreadPoolExecutor-3_0 storage_client.py:440 - delete_client_data() - DEBUG - cleaning up tmp dir: /var/folders/v3/6p7tcx5x7q5cr9c06ky7blvh0000gp/T/tmpdu0443r2
2022-07-29 00:23:42,859 - MainThread file_transfer_agent.py:706 - result() - DEBUG - raise_put_get_error: True, UPLOADED, <class 'str'>, ResultStatus.ERROR, <enum 'ResultStatus'>
2022-07-29 00:23:42,859 - MainThread cursor.py:836 - _init_result_and_meta() - DEBUG - Query result format: json
2022-07-29 00:23:42,859 - MainThread result_batch.py:440 - _parse() - DEBUG - parsing for result batch id: 1
2022-07-29 00:23:42,861 - MainThread cursor.py:625 - execute() - DEBUG - executing SQL/command
2022-07-29 00:23:42,861 - MainThread cursor.py:710 - execute() - INFO - query: [insert into _STAGE_USER.BINLOG_TEST_XUZHENGXI_V2 select column1, column2, column...]
2022-07-29 00:23:42,862 - MainThread connection.py:1252 - _next_sequence_counter() - DEBUG - sequence counter: 4
2022-07-29 00:23:42,862 - MainThread cursor.py:455 - _execute_helper() - DEBUG - Request id: 15a2b33a-5753-40e1-a924-bd52941769ef
2022-07-29 00:23:42,862 - MainThread cursor.py:457 - _execute_helper() - DEBUG - running query [insert into _STAGE_USER.BINLOG_TEST_XUZHENGXI_V2 select column1, column2, column...]
2022-07-29 00:23:42,862 - MainThread cursor.py:466 - _execute_helper() - DEBUG - is_file_transfer: False
2022-07-29 00:23:42,862 - MainThread connection.py:922 - cmd_query() - DEBUG - _cmd_query
2022-07-29 00:23:42,862 - MainThread connection.py:949 - cmd_query() - DEBUG - sql=[insert into _STAGE_USER.BINLOG_TEST_XUZHENGXI_V2 select column1, column2, column...], sequence_id=[4], is_file_transfer=[False]
2022-07-29 00:23:42,863 - MainThread network.py:1138 - _use_requests_session() - DEBUG - Session status for SessionPool 'zcqbucv-aviagames.snowflakecomputing.com', SessionPool 1/1 active sessions
2022-07-29 00:23:42,863 - MainThread network.py:831 - _request_exec_wrapper() - DEBUG - remaining request timeout: None, retry cnt: 1
2022-07-29 00:23:42,863 - MainThread network.py:809 - add_request_guid() - DEBUG - Request guid: f61948cd-7677-4e95-beed-45a533bb7eed
2022-07-29 00:23:42,863 - MainThread network.py:994 - _request_exec() - DEBUG - socket timeout: 60
2022-07-29 00:23:43,199 - MainThread connectionpool.py:461 - _make_request() - DEBUG - https://zcqbucv-aviagames.snowflakecomputing.com:443 "POST /queries/v1/query-request?requestId=15a2b33a-5753-40e1-a924-bd52941769ef&request_guid=f61948cd-7677-4e95-beed-45a533bb7eed HTTP/1.1" 200 373
2022-07-29 00:23:43,200 - MainThread network.py:1023 - _request_exec() - DEBUG - SUCCESS
2022-07-29 00:23:43,200 - MainThread network.py:1144 - _use_requests_session() - DEBUG - Session status for SessionPool 'zcqbucv-aviagames.snowflakecomputing.com', SessionPool 0/1 active sessions
2022-07-29 00:23:43,201 - MainThread network.py:718 - _post_request() - DEBUG - ret[code] = 002049, after post request
2022-07-29 00:23:43,201 - MainThread network.py:740 - _post_request() - DEBUG - Query id: 01a5e977-0004-2723-0008-a20706728616
2022-07-29 00:23:43,201 - MainThread cursor.py:732 - execute() - DEBUG - sfqid: 01a5e977-0004-2723-0008-a20706728616
2022-07-29 00:23:43,201 - MainThread cursor.py:734 - execute() - INFO - query execution done
2022-07-29 00:23:43,201 - MainThread cursor.py:784 - execute() - DEBUG - {'data': {'internalError': False, 'errorCode': '002049', 'age': 0, 'sqlState': '42601', 'queryId': '01a5e977-0004-2723-0008-a20706728616', 'line': -1, 'pos': -1, 'type': 'COMPILATION'}, 'code': '002049', 'message': 'SQL compilation error: error line 1 at position 152\nBind variable ? not set.', 'success': False, 'headers': None}

**Comment by xzxqq1** [comment]
Can someone please reply to this question, thanks a lot

**Comment by frankschmitt** [comment]
I'm running into the same error - I've got a lengthy INSERT statement with lots of . When I run execute_many() with 1 row as input, it succeeds, but with 2 or more rows, it fails.

It seems to be caused by the SELECT in the INSERT statement - I was able to replace it with VALUES(...), and then it works. However, in some circumstances, using SELECT would be preferable (e.g. if you want to use try_to_timestamp_ltz() or similar).
